benchmark/common.js

```diff
@@ -1,205 +1,207 @@
-      else
-        run();
-    });
-  }
-  run();
-};
+  const self = this;
+
+  // If there is only one configuarion don't spawn a process, just run
+  // the benchmark function
+  if (this.queue.length === 1) return this.fn(this.queue[0]);
 
-function parseOpts(options) {
-  // verify that there's an option provided for each of the options
-  // if they're not *all* specified, then we return null.
-  var keys = Object.keys(options);
-  var num = keys.length;
-  var conf = {};
-  for (var i = 2; i < process.argv.length; i++) {
-    var match = process.argv[i].match(/^(.+)=(.*)$/);
-    if (!match || !match[1] || !options[match[1]]) {
-      return null;
-    } else {
-      conf[match[1]] = (match[2].length && isFinite(match[2])
-                        ? +match[2]
-                        : match[2]);
-      num--;
+  (function recursive(queueIndex) {
+    const argv = [];
+    const conf = self.queue[queueIndex];
+    for (const key of Object.keys(conf)) {
+      argv.push(`${key}=${conf[key]}`);
     }
-  }
-  // still go ahead and set whatever WAS set, if it was.
-  if (num !== 0) {
-    Object.keys(conf).forEach(function(k) {
-      options[k] = [conf[k]];
+
+    const child = child_process.fork(require.main.filename, argv);
+    child.on('message', sendResult);
+    child.on('close', function(code) {
+      if (code) {
+        process.exit(code);
+        return;
+      }
+
+      if (queueIndex + 1 < self.queue.length) {
+        recursive(queueIndex + 1);
+      }
     });
-  }
-  return num === 0 ? conf : null;
-}
+  })(0);
+};
 
 Benchmark.prototype.start = function() {
   if (this._started)
     throw new Error('Called start more than once in a single benchmark');
 
   this._started = true;
-  this._start = process.hrtime();
+  this._time = process.hrtime();
 };
 
 Benchmark.prototype.end = function(operations) {
-  var elapsed = process.hrtime(this._start);
+  // get elapsed time now and do error checking later for accuracy.
+  const elapsed = process.hrtime(this._time);
 
-  if (!this._started)
+  if (!this._started) {
     throw new Error('called end without start');
-  if (typeof operations !== 'number')
+  }
+  if (typeof operations !== 'number') {
     throw new Error('called end() without specifying operation count');
+  }
 
-  var time = elapsed[0] + elapsed[1] / 1e9;
-  var rate = operations / time;
-  this.report(rate);
+  const time = elapsed[0] + elapsed[1] / 1e9;
+  const rate = operations / time;
+  this.report(rate, elapsed);
 };
 
-Benchmark.prototype.report = function(value) {
-  var heading = this.getHeading();
-
-  if (outputFormat == 'default')
-    console.log('%s: %s', heading, value.toFixed(5));
-  else if (outputFormat == 'csv')
-    console.log('%s,%s', heading, value.toFixed(5));
+function formatResult(data) {
+  // Construct confiuration string, " A=a, B=b, ..."
+  let conf = '';
+  for (const key of Object.keys(data.conf)) {
+    conf += ' ' + key + '=' + data.conf[key];
```

Comment:
`data.conf[key]` may need to be `JSON.stringify()`ed for strings with control characters (including newlines, etc.), otherwise you'll end up with messed up output. I had to change this with the existing benchmark runner when benchmarking some http header character validation functions.